python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,362 @@
+
+        return order[::-1]
+
+
+class AutoDagNetwork(keras.Model):
+    """DAG model architecture dynamically inferred from model config that maps features -> logits"""
+    INPUTS = "inputs"
+    INPUTS_AS_LIST = "aslist"
+    LAYERS_IDENTIFIER = "layers"
+    LAYER_TYPE = "type"
+    NAME = "name"
+    OP_IDENTIFIER = "op"
+    LAYER_KWARGS = "args"
+
+    def __init__(
+            self,
+            model_config: dict,
+            feature_config: FeatureConfig,
+            file_io: FileIO,
+            **kwargs
+    ):
+        """
+        Initialize a dense neural network layer
+
+        Parameters
+        ----------
+        model_config: dict
+            Dictionary defining the dense architecture spec
+        feature_config: FeatureConfig
+            FeatureConfig defining how each input feature is used in the model
+        file_io: FileIO
+            File input output handler
+        """
+        super().__init__(**kwargs)
+
+        self.file_io: FileIO = file_io
+        self.model_config = model_config
+        self.feature_config = feature_config
+
+        # Get all available layers (including keras layers)
+        self.available_layers = get_layer_subclasses()
+        model_graph = self.define_architecture(model_config)
+        self.execution_order: List[LayerNode] = model_graph.topological_sort()
+        # The line below is important for tensorflow to register the available params for the model
+        # An alternative is to do this in build()
+        # If removed, no layers will be present in the AutoDagNetwork (in the model summary)
+        self.register_layers: List[layers.Layer] = [layer_node.layer for layer_node in self.execution_order if
+                                                    not layer_node.is_input_node]
+        self.file_io.logger.info("Execution order: %s", self.execution_order)
+        self.output_node = model_graph.output_node
+
+    def instantiate_op(self, current_layer_type: str, current_layer_args: Dict) -> layers.Layer:
+        """
+        Create and return an instance of `current_layer_type` with `current_layer_args` params
+        If `current_layer_type` is not found in the set of subclasses of keras.layers.Layer, throws KeyError
+
+        Parameters
+        ----------
+        current_layer_type: str
+            Refers to class name inheriting directly or indirectly from keras.layers.Layer to be instantiated
+        current_layer_args: dict
+            All the arguments from model config which are used as kwargs to current_layer_type instance
+
+        Returns
+        -------
+        keras.layers.Layer
+            Instance of current_layer_type
+        """
+        try:
+            return self.available_layers[current_layer_type](**current_layer_args)
+        except KeyError:
+            self.file_io.logger.error("Layer type: '%s' is not supported or not found in %s",
+                                      current_layer_type, str(self.available_layers))
```

Comment:
@arvindsrikantan Is `available_layers` going to print every single layer under `tf.keras`?

Reply:
Yes, not only tf.keras, but all children classes in the context (example layers from tensorflow_probability are also shown)